core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala

```diff
@@ -901,10 +906,14 @@ trait EntityComponent {
       EntityAndAttributesRawSqlQuery.countReferencesToType(workspaceContext, entityType).map(_.sum)
     }
 
-    def cloneEntitiesToNewWorkspace(sourceWs: UUID, destWs: UUID): WriteAction[(Int, Int)] = {
+    def cloneWorkspace(sourceWs: UUID, destWs: UUID): WriteAction[(Int, Int)] = {
+      copyEntitiesToNewWorkspace(sourceWs, destWs)
+    }
+
+    def copyEntitiesToNewWorkspace(sourceWs: UUID, destWs: UUID, entityRefs: Set[AttributeEntityReference] = Set()): WriteAction[(Int, Int)] = {
       for {
-        entitiesCopiedCount <- CopyEntitiesQuery.copyEntities(sourceWs, destWs)
-        attributesCopiedCount <- CopyEntityAttributesQuery.copyAllAttributes(sourceWs, destWs)
+        entitiesCopiedCount <- CopyEntitiesQuery.copyEntities(sourceWs, destWs, entityRefs)
+        attributesCopiedCount <- CopyEntitiesQuery.copyAttributes(sourceWs, destWs, entityRefs)
       } yield {
         (entitiesCopiedCount, attributesCopiedCount)
       }
```

Comment:
I'm wondering if it makes sense to move the batching of `entityRefs` into here, instead of doing it inside `checkAndCopyEntities`. As-is, this PR is fine (and I wouldn't block its merging based on this comment), but in the future if other code needs to call `copyEntitiesToNewWorkspace`, those code locations would also have to implement their own batching before calling this. Additionally, if you moved the batching here, you have the opportunity to execute all the batches and then sum their results to return an accurate `(Int, Int)` return count. What do you think?

Reply:
I hear you on moving the batching into `copyEntitiesToNewWorkspace.`  It would help callers avoid bugs, but it complicates sharing that code between cloning workspaces and copying entities--at least it does for my rudimentary scala brain.  Here's a first stab at what that code would look like and I'm genuinely unsure about the tradeoffs.  
```
def copyEntitiesToNewWorkspace(sourceWs: UUID, destWs: UUID, entityRefs: Set[AttributeEntityReference] = Set()): WriteAction[(Int, Int)] = {

def copyChunkOfEntitiesOrAllEntities(chunk: Set[AttributeEntityReference] = Set()) = {
        for {
          entitiesCopiedCount <- CopyEntitiesQuery.copyEntities(sourceWs, destWs, chunk)
          attributesCopiedCount <- CopyEntitiesQuery.copyAttributes(sourceWs, destWs, chunk)
        } yield {
          (entitiesCopiedCount, attributesCopiedCount)
        }
      }
      val copyActions = if(entityRefs.isEmpty) Iterator(copyChunkOfEntitiesOrAllEntities()) else entityRefs.grouped(batchSize) map { chunk =>
        copyChunkOfEntitiesOrAllEntities(chunk)
      }
      val allCopies = DBIO.sequence(copyActions)

      allCopies.map { copyActionResults: Iterator[(Int, Int)] =>
        (copyActionResults.map(_._1).sum, copyActionResults.map(_._2).sum)
      }
}
```
I'm wondering if there's a better to get code reuse than using an empty set as a default parameter.

Reply:
I think the problem you're running into here is that `Set.grouped` will return an empty list if the set is empty, correct? And you always need at least one chunk on which to operate, if if that single chunk is the empty set?

So how about ensuring `chunks` has the shape you want, something like:
```scala
val chunks: Iterator[Set[AttributeEntityReference]] = if (entityRefs.nonEmpty) {
  entityRefs.grouped(batchSize)
} else {
  Iterator(Set.empty[AttributeEntityReference])
}
```
or perhaps even better:
```scala
val chunks: Iterator[Set[AttributeEntityReference]] = if (entityRefs.size > batchSize) {
  entityRefs.grouped(batchSize)
} else {
  Iterator(entityRefs)
}
```
and then you can proceed to map over `chunks`, knowing it's correct